pages/blog/graphql-error-handling-with-fp.mdx

```diff
@@ -0,0 +1,640 @@
+
+## Implementing Union types based error handling
+
+Given the following Schema:
+
+```graphql
+type Entity {
+  id: ID!
+  name: String!
+}
+
+type Query {
+  entity(id: ID!, userId: ID!): Entity
+}
+```
+
+If something fails during the execution, we can:
+- throw an Error and use the default GraphQL error mechanism explained above
+- return `null`: this is not satisfactory because we can't know the reason the operation returned `null`
+
+Let's instead implement error handling using Union types with the following modifications:
+
+```graphql
+interface BaseError {
+  message: String!
+}
+
+type InvalidInputError implements BaseError {
+  message: String!
+}
+
+type NotFoundError implements BaseError {
+  message: String!
+}
+
+type UnknownError implements BaseError {
+  message: String!
+}
+
+type NotAllowedError implements BaseError {
+  message: String!
+}
+
+type Entity {
+  id: ID!
+  name: String!
+}
+
+union EntityResult =
+  | Entity
+  | NotFoundError
+  | NotAllowedError
+  | InvalidInputError
+  | UnknownError
+
+type Query {
+  entity(id: ID!, userId: ID!): EntityResult!
+}
+```
+
+We define a `BaseError` interface that all concrete Errors implement.
+Then, we define the query result as a Union type of the happy path (an `Entity`) and the other possible results (`NotFoundError`, `NotAllowedError`, etc.). In this manner, we extensively describe all possible results, with the added benefit that we can make our result type non-nullable (with the `!` character).
+
+For a more detailed explanation, you can read this article from Sasha Solomon: [200 OK! Error Handling in GraphQL](https://sachee.medium.com/200-ok-error-handling-in-graphql-7ec869aec9bc).
+
+## Handle those Schema changes into the GraphQL server
+
+Since we changed the Schema, we also need to change the query, mutation and types resolvers.
+
+For simple cases, we can just return the correct object from the query resolver:
+
+```typescript
+const resolvers = {
+  Query: {
+    entity: async (parent, { id, userId }, context) => {
+      const entity = await context.api.getEntityById(id);
+      const isAllowed = await context.api.isUserAllowedForEntity(id, userId);
+
+      if (!isAllowed) {
+        return {
+          __typename: 'NotAllowedError',
+          message: `User ${userId} not allowed for entity ${id}`,
+        };
+      }
+      if (!entity) {
+        return {
+          __typename: 'NotFoundError',
+          message: `Entity ${id} not found`,
+        };
+      }
+      return {
+        __typename: 'Entity',
+        ...entity,
+      };
+    },
+  },
+};
+```
+
+Since we include the `__typename` and the fields match the Schema type definitions, there is no need to define type resolvers.
+
+For more details on this solution, you can read this article from [Laurin Quast](https://github.com/n1ru4l): [Handling GraphQL errors like a champ with unions and interfaces](https://blog.logrocket.com/handling-graphql-errors-like-a-champ-with-unions-and-interfaces/).
+
```

Comment:
Maybe worth adding here also the possible OneOf solution:
https://github.com/graphql/graphql-spec/pull/825

and how to use it already today: https://github.com/graphql/graphql-spec/pull/825#issuecomment-1039997787

Reply:
## OneOf

If your GraphQL setup supports `@oneOf` (like with Envelop), you can use it for that.

oneOf could completely replace this and makes `__typename` obsolete.

You basically have to decide: do you want an error *or* the payload, or do you want the payload *and* errors. I tend to do the latter, the existence of a payload isn't sufficient to indicate success - you need to request something (the thing you created, a success boolean, whatever) to see if it succeeded. This is the best way, IMO, of giving the user type-safe details of what went wrong whilst still allowing for partial success.

e.g. "saveAndPublish" might successfully save but fail to publish; we might not want to roll both back

Another thing that might be interesting for this is literal enums.

```graphql
enum UserResultType {
  error
  success
}

type UserSuccess {
  type: UserResultType.success
}

type UserError {
  type: UserResultType.error
}

union UserResult = UserSuccess | UserError
```

or you could use an interface

```graphql
type UserSuccess implements UserResult {
  type: UserResultType.success
}

type UserError implements UserResult  {
  type: UserResultType.error
}

interface UserResult {
  type: UserResultType
}
```

and no `__typename` needed and Codegen could then benefit from this.

Reply:
if you don't mind checking it out and if after that you would think it's valuable, I would love it if you do

Reply:
I'll experiment with it on my repo.

Reply:
> If your GraphQL setup supports `@oneOf` (like with Envelop), you can use it for that.
> 
> oneOf could completely replace this and makes `__typename` obsolete.
I've tried the @oneOf directive. It only works on Input type and on Query field (in order to select one of the arguments of the query.

As it is implemented currently, I don't see a way to use it for output Types in order to substitute the use of Union.

Am I missing something?

Reply:
> Another thing that might be interesting for this is literal enums.
> 
> ```graphql
> enum UserResultType {
>   error
>   success
> }
> 
> type UserSuccess {
>   type: UserResultType.success
> }
> 
> type UserError {
>   type: UserResultType.error
> }
> 
> union UserResult = UserSuccess | UserError
> ```
> 
> or you could use an interface
> 
> ```graphql
> type UserSuccess implements UserResult {
>   type: UserResultType.success
> }
> 
> type UserError implements UserResult  {
>   type: UserResultType.error
> }
> 
> interface UserResult {
>   type: UserResultType
> }
> ```
> and no __typename needed and Codegen could then benefit from this.

With this solution, the goal would be to improve the consumption of the query, not the server-side code, right?
e.g. instead of having typeguard client-side like:
```ts
function isGraphqlError(v: unknown): v is BaseError {
  return (v as any)?.__typename.endsWith('Error');
}
```
I would have:
```ts
function isGraphqlError(v: unknown): v is BaseError {
  return (v as any)?.type === ResultType.error;
}
```

Reply:
cc @n1ru4l could you validate?